design/worker-config-drive.md

```diff
@@ -0,0 +1,178 @@
+an ISO image with the data in the right layout. It will then read the
+ISO image into memory and compress it, before writing it to a new
+Secret. It will include a reference to that second Secret with the
+other provisioning instructions (image name, checksum, etc.) it gives
+to the baremetal operator when it allocates a host. The baremetal
+operator will access the secret and pass the contents to Ironic as the
+config drive content.
+
+The base64-encoded compressed ISO is ~8K, and it fits into a Secret
+without issue.
+
+The OpenStack cloud provider uses a value in the provider spec in the
+Machine object to determine the name of this secret. We could do
+something similar in our actuator, which would make it easier to
+support generic Kubernetes as well as OpenShift.
+
+### Risks and Mitigations
+
+We will need to manage the life-cycle of the Secret objects containing
+the ISO images to ensure they are deleted when they are not being
+used. We can set them up so the original Secret is their owner, so the
+ISO is deleted when the original Secret is deleted, or we could create
+one new Secret per Machine and delete them when the Machine is
+deleted. The latter approach will consume more space.
+
+The actuator needs to know about the expectations of the image being
+deployed. That will be easy for us to manage for OpenShift, since we
+will use well-known images there. It will be harder for generic
+Kubernetes because the images may not be known in advance. As future
+work, we can consider having something outside of the actuator build
+the ISO images and associate them with Machine objects (or MachineSet
+objects), to decouple the actuator from the code that manages the
+image format.
+
+## Design Details
+
+### Work Items
+
+- Add a `ConfigDriveRef` field of type `SecretRef` to the
+  `BareMetalHostSpec` structure to hold the location of the Secret. We
+  may want to define a new type to hold all of the provisioning
+  instructions, rather than adding individual fields to the host spec
+  directly.
+- Update the actuator to find the right config drive content and turn
+  it into an ISO stored in a Secret.
+- Update the actuator to pass the ISO Secret to the baremetal operator
+  through the new field in the `BareMetalHostSpec`.
+- Update the baremetal operator to retrieve the config drive content
+  and pass it to Ironic, when it is present.
+
+### Dependencies
+
+This will require work in both the actuator and operator repositories.
+
+### Test Plan
+
+Unit tests for the manipulation of the Secrets and creating the ISO
+image.
+
+Manual or automated integration tests for ensuring the config drive
+content is applied to the server during provisioning.
+
+### Upgrade / Downgrade Strategy
+
+The config drive contents shouldn't change as part of an upgrade,
+since they are a minimal set of instructions to tell the host how to
+reach the cluster, which then provides more detailed instructions.
```

Comment:
These should be "one time" pieces of data per each node being deployed, so they really should be able to be disregarded or abandoned after the deployment has been triggered.